### PR TITLE
feat(apm): enrichment engine - shared primitives + enricher query/format

### PR DESCRIPTION
**Pull request by jonmcwest**
## Problem

Editors and AI agents have no visibility into production latency data for the code they're reading. This adds the foundational layer for APM (tracing-span) enrichment, enabling per-line span statistics from PostHog to be fetched and rendered as inline comments alongside source code.

## Changes

**Shared types and utilities (`@posthog/shared`)**
- Added `SpanLineStat`, `SymbolStatsRow`, `SymbolStatsPeriod`, `SourceSymbol`, `ApmWindow`, and related interfaces as the cross-package boundary types for APM enrichment.
- Added `APM_STATS_WINDOW` (defaulting to `24h`) as the single source of truth for the query window, so the date range, label, and comparison label all derive from one value.
- Added `formatMs` (compact millisecond formatting) and `formatPercentDelta` (period-over-period % change, suppressing sub-1% noise and non-finite values).
- Added `APM_LANG_BY_EXT`, `apmLangForFile`, and `fileExtension` to determine comment style eligibility from a file path.
- Added `APM_ENRICHMENT_FLAG` feature flag constant.

**Query builder and result mapper (`packages/enricher`)**
- Added `buildSymbolStatsQuery` in `apm-breakdown.ts`, which constructs the `TraceSpansSymbolStatsQuery` request body. Omits `symbols` for per-line mode; includes them for per-symbol rollup. Defaults the window to `APM_STATS_WINDOW`.
- Added `mapSymbolStatsResults` to convert raw server rows (nanosecond durations, snake_case fields) into `SpanLineStat` client shapes (milliseconds, camelCase), preserving server ordering and passing through server-computed deltas.
- Added `getApmLineStats` to `PostHogApi`, which posts to `/tracing/spans/symbol-stats/` and returns mapped `SpanLineStat[]`.

**Inline comment formatter (`packages/enricher`)**
- Added `formatApmInlineComments` in `apm-comment-formatter.ts`, which annotates source lines with APM metrics (p50/p95/p99, span count, error count, period-over-period deltas) as inline comments. Handles CRLF line endings, out-of-range line numbers, and per-language comment prefixes (`#` vs `//`).
- Extracted `commentPrefix` into a shared `comment-style.ts` module so the `#`-vs-`//` rule is no longer duplicated between the event/flag and APM formatters.
- Exported `formatApmInlineComments` from the enricher's public index.

## How did you test this?

Unit tests were added and run for all new logic:
- `apm-breakdown.test.ts`: covers `buildSymbolStatsQuery` (line mode, symbol mode, window defaults, empty symbols) and `mapSymbolStatsResults` (ns→ms conversion, delta passthrough, ordering, empty input).
- `apm-comment-formatter.test.ts`: covers comment injection at the correct line, drill-in hint format, language-specific prefixes, error count visibility, out-of-range stats, count unit promotion (k/M), p99 and window label rendering, period-over-period delta formatting, and CRLF preservation.
- `apm-enrichment.test.ts`: covers `formatMs` rounding behaviour and `formatPercentDelta` edge cases (null, non-finite, sub-1% suppression, symmetric rounding).

## Automatic notifications

- [ ] Publish to changelog?
- [ ] Alert Sales and Marketing teams?

### CODE REVIEW

**Comment by jonmcwest** [comment]
* **#2839** <a href="https://app.graphite.com/github/pr/PostHog/code/2839?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2838** <a href="https://app.graphite.com/github/pr/PostHog/code/2838?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2836** <a href="https://app.graphite.com/github/pr/PostHog/code/2836?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2776** <a href="https://app.graphite.com/github/pr/PostHog/code/2776?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/PostHog/code/2776?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: packages/shared/src/apm-enrichment.ts [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+ * requested symbol range (symbol mode). Flat fields are the current period.
+ */
+export interface SymbolStatsRow extends SymbolStatsPeriod {
+  line: number;
+  /** Symbol mode only. */
+  name?: string;
+  /** Symbol mode only. */
+  end_line?: number;
+  /** Still returned but no longer read by the client; slated for removal. */
+  previous: SymbolStatsPeriod;
```
Comment (jonmcwest) [inline packages/shared/src/apm-enrichment.ts]: this is 'slated for removal' but wasn't removed?

File: packages/shared/src/apm-enrichment.ts [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+  p99_busy_nano: number;
+}
+
+/**
+ * One bucket of the symbol-stats response: per source line (line mode) or per
+ * requested symbol range (symbol mode). Flat fields are the current period.
+ */
+export interface SymbolStatsRow extends SymbolStatsPeriod {
+  line: number;
+  /** Symbol mode only. */
```
Comment (jonmcwest) [inline packages/shared/src/apm-enrichment.ts]: "Symbol mode" is a smell, this should probably be a seperate type/interface

File: packages/shared/src/apm-enrichment.ts [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+  count_pct_change: number | null;
+  p50_duration_pct_change: number | null;
+  p95_duration_pct_change: number | null;
+  p99_duration_pct_change: number | null;
+  error_rate_pct_change: number | null;
+}
+
+export type SymbolStatsGranularity = "line" | "symbol";
+
+/** Format a millisecond value compactly: one decimal under 10ms, else rounded. */
```
Comment (jonmcwest) [inline packages/shared/src/apm-enrichment.ts]: the code doesnt need a comment it speaks for itself

File: packages/shared/src/apm-enrichment.ts [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+export function formatPercentDelta(
+  pct: number | null | undefined,
+): string | null {
+  if (pct == null || !Number.isFinite(pct)) return null;
+  // Round half away from zero so equal-magnitude deltas render symmetrically:
+  // Math.round sends -0.5 to -0 (suppressed) while +0.5 becomes +1 (shown).
+  const rounded = pct < 0 ? -Math.round(-pct) : Math.round(pct);
+  if (rounded === 0) return null;
+  return `${rounded > 0 ? "+" : ""}${rounded}%`;
+}
```
Comment (jonmcwest) [inline packages/shared/src/apm-enrichment.ts:122]: We are getting the percent deltas from the API now, please review all of the formatPercentDelta implementations/tests in this PR because i'm not sure they are necessary

File: packages/shared/src/apm-enrichment.ts (hunk 1 above)
Comment (jonmcwest) [reply packages/shared/src/apm-enrichment.ts]: Removed — the client no longer reads `previous` (the deltas come from the API now), so it is dropped from the type.

File: packages/shared/src/apm-enrichment.ts (hunk 2 above)
Comment (jonmcwest) [reply packages/shared/src/apm-enrichment.ts]: Stripped `name?`/`end_line?` from the line-mode row so it is pure line-mode; the symbol-mode row type lands in PR4 where it is actually consumed.

File: packages/shared/src/apm-enrichment.ts (hunk 3 above)
Comment (jonmcwest) [reply packages/shared/src/apm-enrichment.ts]: Removed.

File: packages/shared/src/apm-enrichment.ts (hunk 4 above)
Comment (jonmcwest) [reply packages/shared/src/apm-enrichment.ts:122]: Kept — this is the display formatter for the raw API deltas (we removed the computation, `pctChange`). It rounds/signs and suppresses sub-1% noise, and is shared by the editor popover + agent comments. The only now-defensive bit is the non-finite guard (the API guarantees finite-or-null); happy to drop that line if you prefer it leaner.

**Comment by charlesvien** [comment]
Closing stale PRs ahead of Friday's monorepo migration. If this is a mistake, rebase and reopen for review.
